daemon/start.go

```diff
@@ -11,79 +10,95 @@ import (
 	"google.golang.org/grpc"
 
 	"github.com/Sirupsen/logrus"
-	apierrors "github.com/docker/docker/api/errors"
 	"github.com/docker/docker/api/types"
 	containertypes "github.com/docker/docker/api/types/container"
 	"github.com/docker/docker/container"
+	"github.com/pkg/errors"
 )
 
 // ContainerStart starts a container.
 func (daemon *Daemon) ContainerStart(name string, hostConfig *containertypes.HostConfig, checkpoint string, checkpointDir string) error {
 	if checkpoint != "" && !daemon.HasExperimental() {
-		return apierrors.NewBadRequestError(fmt.Errorf("checkpoint is only supported in experimental mode"))
+		return validationError{fmt.Errorf("checkpoint is only supported in experimental mode")}
 	}
 
 	container, err := daemon.GetContainer(name)
 	if err != nil {
 		return err
 	}
 
-	if container.IsPaused() {
-		return fmt.Errorf("Cannot start a paused container, try unpause instead.")
-	}
+	err = func() error {
+		container.Lock()
```

Comment:
Was the lack of locking before a bug? Is it necessary even with the "containers in memdb" changes? (I'm not sure offhand whether `GetContainer` returns copies from that store or actual values)